### DISCUSSION TITLE
Use a local quote or fortunes file and user script for daily notes

### DISCUSSION THREAD

**Discussion by kmaris** [SilentVoid13/Templater#1415] [Templates Showcase]
Here's a small user script for a [Daily Notes](https://help.obsidian.md/Plugins/Daily+notes) template.

The idea is to pull a quote from a quotes-filled vault file. I use a BSD-style quote file [found here](http://bxr.su/OpenBSD/games/fortune/datfiles/fortunes). The quote functionality in Templater did not work for me, as the upstream quote api seems flakey. This will also work with no connectivity like if you're on an airplane.

Download a file, or look at how each quote is delimited by lines containing only the '%' character. This script assumes as such, so if you roll your own you _must_ follow that convention for this script to work. Or tweak my script :)

I assume that:

- You have setup the Templater and Daily Notes plugins, and you are using a template for your daily notes.
- You have a quote file ready-to-go, and inside your vault.

Once you have those, you'll want to create a user script in your vault, as an example I keep mine in `assets/templater-plugin/scripts/fortune.js`. That path is what will be in your templater user scripts plugin settings.

The content of `fortune.js` should look roughly like:

```javascript
const nodefs = require('node:fs');

function fortune(fortunes_path) {
    const fp = app.vault.adapter.getFullRealPath(fortunes_path);
    const data = nodefs.readFileSync(fp, 'utf8');
    const lines = data.split('\n');

    // Pick a random quote in the file
    var randi = Math.floor(Math.random()*lines.length);
    var start = lines.findLastIndex((el, i) => el == '%' && i <= randi);
    if (start == -1) {
        start = 0;
    }
    if (start == randi) {
        randi++;
    }
    var end = lines.findIndex((el, i) => el == '%' && i >= randi);
    if (end == -1) {
        end--;
    }
    return lines.slice(start+1, end).join('\n');
}
module.exports = fortune;
```

> [!IMPORTANT]
> Please note, I am not a javascript developer. So the code is probably not very good but it only runs once a day, quick enough, and I'm fine with that. Feel free to send me any optimizations or add your own!

After you have the script, go ahead and use it in your daily note template(I put my quote in a code block, as the formatting is sometimes kinda rough). You need to give the user script function the path to your quotes file, like so:

~~~
```
<% tp.user.fortune('assets/fortunes.txt') %>
```
~~~

And voila, hopefully you'll have a quote in your daily note.

<img width="730" alt="Screenshot 2024-07-15 at 3 06 41 PM" src="https://github.com/user-attachments/assets/000049ec-c399-44c6-8957-8623d2278f77">

**Comment by jowodo** [comment]
thanks a bunch for this template!
It helped me a lot to adapt the script to select a random line from a markdown file: 
```javascript
// inspired by https://github.com/SilentVoid13/Templater/discussions/1415
const nodefs = require('node:fs');

function randomLine(filePath) {
    const fp = app.vault.adapter.getFullRealPath(filePath);
    const data = nodefs.readFileSync(fp, 'utf8');
    const lines = data.split('\n');
    const nonEmptyLines = lines.filter(Boolean);

    // Pick a random quote in the file
    var randi = Math.floor(Math.random()*nonEmptyLines.length);
return nonEmptyLines[randi];
}

module.exports = randomLine;
```

**Reply by jowodo** [reply]
I managed to boil it down to a single tag without the need for a separate user script! 
Inspired by [AdriaanRol's snippet](https://github.com/SilentVoid13/Templater/issues/586#issuecomment-1874225575): 
```javascript
<%*
fileName = "reflection-questions";
file = tp.file.find_tfile(fileName); 
tR+= ""
if ( file ) {
    fileContent = await tp.file.include(file);
    lines = fileContent.split('\n');
    nonEmptyLines = lines.filter(Boolean);
    let randomIndex = Math.floor(Math.random() * nonEmptyLines.length);
    tR+= nonEmptyLines[randomIndex];
} else {
    tR+= "input file does not exist!"
}
%>
```
I'm pretty happy with this solution because:
1. it uses the internal `Templater` function `tp.file.include()` to read the file 
2. the code is not spread across multiple files
3. finding of files works without path (and `.md` is assumed, but other extensions (e.g. `.txt` work too)

**Comment by kmaris** [comment]
@jowodo Hah I love your version, when I made mine I told myself to look into using actual built in libs but clearly never got around to it. Yours is much cleaner, thank you for posting it. Glad the template helped :smiley: 

Edit: although now I’m noticing you’re pulling individual lines as opposed to the quote file I used which has quotes delimited by ‘%’ so there can be multi-line quotes (such as in my example pic).
